src/algebra/module/zlattice.lean

```diff
@@ -0,0 +1,205 @@
+* As `L : add_subgroup E` with the additional properties:
+  `(hd : ∀ r : ℝ, (L ∩ (metric.closed_ball 0 r)).finite)`, that is `L` is discrete
+  `(hs : submodule.span ℝ (L : set E) = ⊤)`, that is `L` spans `E` over `K`.
+
+## Main result
+* `zspan.is_add_fundamental_domain`: for `L : submodule.span ℤ (set.range b)` a ℤ-lattice, proves
+ that the set defined by `zspan.fundamental_domain` is indeed a fundamental domain.
+
+-/
+
+open_locale big_operators classical
+
+noncomputable theory
+
+namespace zspan
+
+open measure_theory measurable_set submodule
+
+variables {E ι : Type*}
+
+section normed_lattice_field
+
+variables {K : Type*} [normed_lattice_field K]
+variables [normed_add_comm_group E] [normed_space K E]
+variables (b : basis ι K E)
+
+/-- The fundamental domain of the ℤ-lattice spanned by `b`. See `zspan.is_add_fundamental_domain`
+for the proof that it is the fundamental domain. -/
+def fundamental_domain : set E := { m | ∀ i, b.repr m i ∈ set.Ico (0 : K) 1 }
+
+variables [floor_ring K]
+
+section fintype
+
+variable [fintype ι]
+
+/-- The map that sends a vector of `E` to the element of the ℤ-lattice spanned by `b` obtained
+by rounding down its coordinates on the basis `b`. -/
+def floor (m : E) : span ℤ (set.range b) := ∑ i, ⌊b.repr m i⌋ • b.restrict_scalars ℤ i
+
+@[simp]
+lemma repr_floor_apply (m : E) (i : ι) :
+  b.repr (floor b m) i = ⌊b.repr m i⌋ :=
+by simp only [floor, zsmul_eq_smul_cast K, b.repr.map_smul, finsupp.single_apply,
+  finset.sum_apply', basis.repr_self, finsupp.smul_single', mul_one, finset.sum_ite_eq',
+  finset.mem_univ, if_true, coe_sum, coe_smul_of_tower, basis.restrict_scalars_apply,
+  linear_equiv.map_sum]
+
+/-- The map that sends a vector `E` to the fundamental domain of the lattice,
+see `zspan.fract_mem_fundamental_domain`. -/
+def fract : E → E := λ m, m - floor b m
+
+lemma fract_apply (m : E) : fract b m = m - floor b m := rfl
+
+@[simp]
+lemma repr_fract_apply (m : E) (i : ι):
+  b.repr (fract b m) i = int.fract (b.repr m i) :=
+by rw [fract, map_sub, finsupp.coe_sub, pi.sub_apply, repr_floor_apply, int.fract]
+
+@[simp]
+lemma fract_zspan_add (m : E) {v : E} (h : v ∈ span ℤ (set.range b)) :
+  fract b (v + m) = fract b m :=
+begin
+  refine (basis.ext_elem_iff b).mpr (λ i, _),
+  simp_rw [repr_fract_apply, int.fract_eq_fract],
+  use (b.restrict_scalars ℤ).repr ⟨v, h⟩ i,
+  rw [map_add, finsupp.coe_add, pi.add_apply, add_tsub_cancel_right,
+    ← (eq_int_cast (algebra_map ℤ K) _), basis.restrict_scalars_repr_apply, coe_mk],
+end
+
+variable {b}
+
+lemma mem_fundamental_domain {x : E} :
+  x ∈ fundamental_domain b ↔ fract b x = x :=
+by simp only [basis.ext_elem_iff b, fundamental_domain, set.mem_Ico, set.mem_set_of_eq,
+  repr_fract_apply, int.fract_eq_self]
+
+variable (b)
+
+lemma fract_mem_fundamental_domain (x : E) :
+  fract b x ∈ fundamental_domain b :=
+by simp only [mem_fundamental_domain, basis.ext_elem_iff b, repr_fract_apply, int.fract_fract,
+  eq_self_iff_true, implies_true_iff]
```

Comment:
This proof is going too far down because you haven't proved `fract_idem : fract b (fract b x) = fract b x`

Reply:
Ok, so I proved the result and it used it to simplify the proof of `fract_mem_fundamental_domain`. I called it `fract_fract` though to match the `int` lemma name ;)